libsel4/tools/generator_rust.py

```diff
@@ -0,0 +1,604 @@
+#
+# SPDX-License-Identifier: BSD-2-Clause
+#
+
+from generator import (
+    MAX_MESSAGE_LENGTH,
+    Generator,
+    is_result_struct_required,
+    struct_members,
+    get_parameter_positions
+)
+
+
+class Generator_Rust(Generator):
+    @staticmethod
+    def translate_type(name):
+        type_trans = {
+            "int": "isize",
+            "seL4_Uint8": "u8",
+            "seL4_Uint16": "u16",
+            "seL4_Uint32": "u32",
+            "seL4_Uint64": "u64",
+            "seL4_Bool": "u8",
+            "seL4_CapData_t": "seL4_CapData",
+            "seL4_PrioProps_t": "seL4_PrioProps",
+            "seL4_CapRights_t": "seL4_CapRights",
+            "seL4_RISCV_Page_GetAddress_t": "seL4_RISCV_Page_GetAddress",
+        }
+        if name in type_trans:
+            return type_trans[name]
+        else:
+            return name
+
+    @staticmethod
+    def translate_expr(name):
+        # TODO there are cases which implement this in-line, refactor to use this function instead
+        #      or get rid of this
+        if name == "type":
+            return "type_"
+        return name
+
+    TYPES = {
+        8:  "u8",
+        16: "u16",
+        32: "u32",
+        64: "u64"
+    }
+
+    class Type(object):
+        """
+        This class represents a Rust type (such as an 'u32', structure or
+        pointer.
+        """
+
+        def __init__(self, name, size_bits, wordsize, *, native_size_bits=None):
+            """
+            Define a new type, named 'name' that is 'size_bits' bits
+            long.
+            """
+
+            self.name = name
+            self.size_bits = size_bits
+            self.wordsize = wordsize
+            self.double_word = (size_bits == wordsize*2)
+
+            #
+            # Store the number of bits Rust will use for this type
+            # in its native unpacked form.
+            if native_size_bits:
+                self.native_size_bits = native_size_bits
+            else:
+                self.native_size_bits = size_bits
+
+        def pass_by_reference(self):
+            return self.size_bits > self.wordsize and not self.double_word
+
+        def render_parameter_name(self, name):
+            """
+            Return a string of Rust code that would be used in a function
+            parameter declaration.
+            """
+            if name == "type":
+                name = "type_"
+            return "%s: %s" % (name, Generator_Rust.translate_type(self.name))
+
+        def pointer(self, mutable):
```

Comment:
Maybe it make es code a bit easier to understand if there are two functions here
- `asMutablePointerType()`
- `asImmutablePointerType()`